### ISSUE TITLE
Editor Setting assets not loading from disk

### ISSUE THREAD

**Issue by AntonC9018** [hextantstudios/com.hextantstudios.utilities#2]
**Version used:**
Unity 2020.3.12f1

**Summary:**
Custom user editor settings do save the asset file with the current settings on disk, but they are unable to read it from there when the setting is requested for the first time.

**Steps to reproduce:**
1. Clone [the example repo](https://github.com/AntonC9018/HextantIssue) I have created to showcase this. Your repo is included as a submodule of my fork.
```
git clone https://github.com/AntonC9018/HextantIssue --recursive
```
2. Open the folder in Unity. Observe successful initialization.
3. Hit `Do Test/Set to false`. This will print the current default value of the boolean value added via your `Settings` API, which is initially `true`, and set it to `false`. You can check it by hitting `Do Test/Print` and observe it printing `False`.
4. Close Unity editor.
5. See if the setting has been saved properly. Open the asset file at `Assets\Settings\Editor\User\HextantIssue\TestSettings.asset`, and make sure the entry `_boolean: 0` is present.
6. Open the project in Unity again. Hit `Do Test/Print` and observe it printing `True` as the current value of `_boolean`.
7. Open the asset file at `Assets\Settings\Editor\User\HextantIssue\TestSettings.asset`, and observe the entry `_boolean: 1` being there.

This means the existing asset file had not been loaded, instead, a new one has been generated, which has substituted the existing one on disk.

**Code**

I'm going to place the essential code here too. This snippet introduces the boolean setting and the two menu items for the editor.

```C#

using Hextant;
using Hextant.Editor;
using UnityEditor;
using UnityEngine;

namespace Test
{
    [Settings(SettingsUsage.EditorUser, "A boolean setting")]
    public sealed class TestSettings : Settings<TestSettings>
    {
        public bool Boolean { get => _boolean; set => Set(ref _boolean, value); }
        [SerializeField, Tooltip("A boolean value, true by default")] 
        private bool _boolean = true;

        [SettingsProvider]
        static SettingsProvider GetSettingsProvider() => instance.GetSettingsProvider();
    }

    [InitializeOnLoad]
    public class EditorTest
    {
        [MenuItem("Do Test/Print")]
        public static void Print()
        {
            Debug.Log(TestSettings.instance.Boolean);
        }

        [MenuItem("Do Test/Set to false")]
        public static void SetToFalse()
        {
            Debug.Log(TestSettings.instance.Boolean);
            TestSettings.instance.Boolean = false;
        }
    }
}
```

**My guess**

The problem is most likely [here](https://github.com/AntonC9018/com.hextantstudios.utilities/blob/93a166a9c1b75a80cc2124bfb961abe320428af6/Runtime/Settings/Settings.cs#L49-L50). 

If I change this

```C#
else _instance = AssetDatabase.LoadAssetAtPath<T>( path );
```

to this:

```C#
else 
{
    _instance = AssetDatabase.LoadAssetAtPath<T>( path );
    Debug.Log($"Found at {path}? : {!(_instance is null)}");
}
```

I can see it printing `False` the first time the setting is tried to get loaded. Then it obviously goes on and creates a new asset with the default value of `true`. On subsequent domain reloads, the asset would successfully load at this line, printing `True`, however not on editor restart, which makes me think it somehow **only looks at cached settings.**

Please inform if I'm missing a crucial usage detail that would make it load the setting from the path into cache at editor start. I believe, though, this should happen automatically.

**Comment by krisrok** [comment]
Not an issue with this project per se but a Unity quirk: Unity unfortunatley needs the class name to match the file name for the serialization to work correctly for MonoBehaviours and ScriptableObjects (and possibly everything else inheriting from UnityEngine.Object).

You can see it's not working correctly and will likely fail when selecting the asset in the inspector: It cannot find the corresponding script.
Also, what I saw when experimenting: For the AssetDatabase those 'illegal' assets are invisible, that's why LoadAssetAtPath does not work. AssetDatabase.FindAssets("t:Test.TestSettings") also fails as long as the filename does not match.

TLDR: Rename Test.cs to TestSettings.cs and you're all set. Be sure to delete and recreate the asset after the change though.

**Comment by AntonC9018** [comment]
After refactoring the settings class into a separate file it seems to be working correctly. Thank you.

I think mentioning this in the documentation may be a good idea also.

**Comment by Hextant** [comment]
Sorry for the delay replying. Thanks @krisrok for pointing out the issue. I run into that every now and then, and it can be very annoying. I'll update the doc (blog article) and add a note at the top of `Settings.cs`.

**Comment by krisrok** [comment]
In my wip branch I‘ve added a check for this:
https://github.com/krisrok/com.hextantstudios.utilities/blob/7da95e386e7df3c0be0bb388360752d2c3a20121/Runtime/Settings/Settings.cs#L99

**Comment by Hextant** [comment]
Thanks @krisrok! I added a similar check here as well and just changed the `return null` checks to throw exceptions since there should never really be a valid case where it returns `null` and it prevents following `NullReferenceExceptions`...
